File: Scripts/TeleportTouch.cs
using UnityEngine;
using System.Collections;
using System.Text;

public class TeleportTouch : MonoBehaviour {
    public int teleportID; // Unique ID for this instance
    public int targetDestinationID; // Destination teleport ID (linked via index)
	public float justUsed = 0f; // save
	public bool touchEnabled = true;
	
    private static TeleportTouch[] allTeleportTouches = new TeleportTouch[8];
	private static StringBuilder s1 = new StringBuilder();
	
	public void Awake() {
		if (teleportID > allTeleportTouches.Length || teleportID < 0) { Destroy(this.gameObject); return; }
		
        allTeleportTouches[teleportID] = this;
    }

	void  OnTriggerEnter ( Collider col  ) {
		if (!touchEnabled) return;
		if (col == null) return;
		if (col.gameObject == null) return;

		if (col.gameObject.CompareTag("Player")) {
			HealthManager hm = Utils.GetMainHealthManager(col.gameObject);
			if (hm != null) {
				if (hm.health > 0f && justUsed < Eng_Global->pauseRelativeTime) {
					Eng_UI->teleportFX.SetActive(true);
					TeleportTouch tt = allTeleportTouches[targetDestinationID];
					if (tt != null) {
						col.Eng_Global->instances[i].position = tt.Eng_Global->instances[i].position; // Do it!
						tt.justUsed = Eng_Global->pauseRelativeTime + 1.0f;
					}
					
					Utils.PlayUIOneShotSavable(106);
				}
			}
		}
	}
}

File: Scripts/teleporttouch.c
#include "mod.h"

static uint16_t TeleportTouch_allTeleportTouches[8];
static bool TeleportTouch_initialized;

void TeleportTouchInitAfterLoad(uint16_t self) {
    Entity* e = &Eng_Global->instances[self];
    if (!TeleportTouch_initialized) { for (uint8_t i = 0; i < 8; i++) TeleportTouch_allTeleportTouches[i] = UINT16_MAX; TeleportTouch_initialized = true; }
    if (e->teleportID >= 8) { DeleteInstance(self); return; }
    TeleportTouch_allTeleportTouches[e->teleportID] = self;
}

void TeleportTouchOnTriggerEnter(uint16_t self, uint16_t other) {
    Entity* e = &Eng_Global->instances[self];
    Entity* player = &Eng_Global->instances[PLAYER1];
    if (!e->touchEnabled || other != PLAYER1) return;
    if (player->health <= 0.0f || e->justUsed >= Eng_Global->pauseRelativeTime) return;
    uint16_t dest = e->targetDestinationID < 8 ? TeleportTouch_allTeleportTouches[e->targetDestinationID] : UINT16_MAX;
    if (dest == UINT16_MAX) return;
    player->position = Eng_Global->instances[dest].position;
    Eng_Global->instances[dest].justUsed = Eng_Global->pauseRelativeTime + 1.0;
    play_wav(sounds[106],1.0f,Eng_Global->instances[dest].position,false);
}
